File: bifrost-rs/crates/bifrost-runtime/src/socket.rs
//! `BifrostSocket` trait — the minimal surface the runtime needs from
//! a transport-owned WebSocket handle.

use async_trait::async_trait;

/// Numeric ready states matching JS `WebSocket.readyState`.
pub mod state {
    pub const CONNECTING: u8 = 0;
    pub const OPEN: u8 = 1;
    pub const CLOSING: u8 = 2;
    pub const CLOSED: u8 = 3;
}

pub use state as SocketState;

/// Minimal abstraction over a WebSocket. Transports provide a concrete
/// `impl BifrostSocket` that the runtime uses for broadcast, auth
/// result delivery, etc.
#[async_trait]
pub trait BifrostSocket: Send + Sync {
    /// Current connection state — one of the [`state`] constants.
    fn ready_state(&self) -> u8;

    /// Send a text frame. MUST be a no-op if the socket is not open.
    async fn send(&self, data: String);

    /// Close the socket; idempotent.
    async fn close(&self);

    /// Stable identity used for room membership and originator-exclusion.
    fn id(&self) -> u64;
}

File: typeferry-py/src/typeferry/server/socket.py
"""Framework-agnostic socket interface.

Mirrors ``TypeFerrySocket`` and ``SocketState`` from
``typeferry-ts/src/server/types.ts``. The concrete ASGI implementation
adapts Starlette's ``WebSocket`` to this protocol.
"""

from __future__ import annotations

from typing import Any, Protocol


class SocketState:
    """Numeric WebSocket ready states (mirrors JS ``WebSocket.readyState``)."""

    CONNECTING = 0
    OPEN = 1
    CLOSING = 2
    CLOSED = 3


class TypeFerrySocket(Protocol):
    """Minimal socket surface used by the server runtime.

    Concrete implementations MUST expose ``readyState`` as an int
    matching :class:`SocketState`, and ``send`` as an awaitable-or-sync
    method accepting a text frame.
    """

    readyState: int

    def send(self, data: str) -> Any: ...

    def close(self) -> Any: ...
